File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/sessions.rb
# frozen_string_literal: true

require 'json'
require_relative '../errors'
require_relative '../database'
require_relative 'campaigns'

module Handlers
  # Campaign session scheduling: schedule sessions, record attendance, and
  # look up the next upcoming session by start time.
  module Sessions
    module_function

    def find_session(campaign_id, session_id)
      row = Database.query(<<~SQL).first
        SELECT id, campaign_id, starts_at, duration_minutes, agenda_json, present_json, absent_json
        FROM campaign_sessions
        WHERE campaign_id = #{Database.escape(campaign_id)} AND id = #{Database.escape(session_id)};
      SQL
      raise HttpError.new(404, 'unknown session id') unless row

      row
    end

    def create_session(campaign_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)

      id = body['id']
      starts_at = body['starts_at']
      duration_minutes = body['duration_minutes']
      agenda = body['agenda']

      raise HttpError.new(400, 'id must be a string') unless id.is_a?(String) && !id.empty?
      raise HttpError.new(400, 'starts_at must be a string') unless starts_at.is_a?(String) && !starts_at.empty?
      raise HttpError.new(400, 'duration_minutes must be a positive integer') unless duration_minutes.is_a?(Integer) && duration_minutes.positive?
      raise HttpError.new(400, 'agenda must be an array of strings') unless agenda.is_a?(Array) && agenda.all? { |a| a.is_a?(String) }
      raise HttpError.new(409, 'id already exists') unless Database.query("SELECT id FROM campaign_sessions WHERE id = #{Database.escape(id)};").empty?

      Database.exec(<<~SQL)
        INSERT INTO campaign_sessions (id, campaign_id, starts_at, duration_minutes, agenda_json, present_json, absent_json)
        VALUES (
          #{Database.escape(id)},
          #{Database.escape(campaign_id)},
          #{Database.escape(starts_at)},
          #{Database.int(duration_minutes)},
          #{Database.escape(JSON.generate(agenda))},
          #{Database.escape(JSON.generate([]))},
          #{Database.escape(JSON.generate([]))}
        );
      SQL

      [201, {
        id: id,
        starts_at: starts_at,
        duration_minutes: duration_minutes,
        agenda_count: agenda.size
      }]
    end

    def record_attendance(campaign_id, session_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)
      find_session(campaign_id, session_id)

      present = body['present']
      absent = body['absent']

      raise HttpError.new(400, 'present must be an array of strings') unless present.is_a?(Array) && present.all? { |c| c.is_a?(String) }
      raise HttpError.new(400, 'absent must be an array of strings') unless absent.is_a?(Array) && absent.all? { |c| c.is_a?(String) }

      Database.exec(<<~SQL)
        UPDATE campaign_sessions
        SET present_json = #{Database.escape(JSON.generate(present))},
            absent_json = #{Database.escape(JSON.generate(absent))}
        WHERE campaign_id = #{Database.escape(campaign_id)} AND id = #{Database.escape(session_id)};
      SQL

      [200, {
        session_id: session_id,
        present_count: present.size,
        absent_count: absent.size
      }]
    end

    def next_session(campaign_id, _body)
      Handlers::Campaigns.find_campaign(campaign_id)

      rows = Database.query(<<~SQL)
        SELECT id, starts_at, agenda_json
        FROM campaign_sessions
        WHERE campaign_id = #{Database.escape(campaign_id)}
        ORDER BY starts_at ASC;
      SQL
      raise HttpError.new(404, 'no sessions scheduled') if rows.empty?

      row = rows.first
      agenda = JSON.parse(row['agenda_json'])

      [200, {
        id: row['id'],
        starts_at: row['starts_at'],
        agenda_count: agenda.size
      }]
    end
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/sessions.ts
/** Campaign session scheduling: scheduling, attendance, and next-session lookup. */

import { getDb } from '../db.ts';
import type { ApiResult, JsonValue } from '../types.ts';
import { isValidIntInRange } from '../validation.ts';

function isIsoTimestamp(value: unknown): value is string {
  if (typeof value !== 'string' || value.length === 0) return false;
  const parsed = Date.parse(value);
  return !Number.isNaN(parsed);
}

interface CampaignSessionRow {
  id: string;
  starts_at: string;
  duration_minutes: number;
  agenda_json: string;
}

export function scheduleSession(campaignId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const id = body.id;
  if (typeof id !== 'string' || id.length === 0) {
    return { status: 400, body: { error: 'id must be a non-empty string' } };
  }

  const startsAt = body.starts_at;
  if (!isIsoTimestamp(startsAt)) {
    return { status: 400, body: { error: 'starts_at must be a valid ISO 8601 timestamp' } };
  }

  const durationMinutes = body.duration_minutes;
  if (!isValidIntInRange(durationMinutes, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'duration_minutes must be a positive integer' } };
  }

  const agenda = body.agenda;
  if (!Array.isArray(agenda) || !agenda.every((entry) => typeof entry === 'string')) {
    return { status: 400, body: { error: 'agenda must be an array of strings' } };
  }

  const existing = db
    .prepare('SELECT id FROM campaign_sessions WHERE campaign_id = ? AND id = ?')
    .get(campaignId, id);
  if (existing) {
    return { status: 409, body: { error: 'session id already exists' } };
  }

  db.prepare(
    'INSERT INTO campaign_sessions (campaign_id, id, starts_at, duration_minutes, agenda_json) VALUES (?, ?, ?, ?, ?)',
  ).run(campaignId, id, startsAt as string, durationMinutes as number, JSON.stringify(agenda));

  return {
    status: 201,
    body: {
      id,
      starts_at: startsAt,
      duration_minutes: durationMinutes,
      agenda_count: (agenda as string[]).length,
    },
  };
}

function isStringArray(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((entry) => typeof entry === 'string');
}

export function recordAttendance(campaignId: string, sessionId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const session = db
    .prepare('SELECT id FROM campaign_sessions WHERE campaign_id = ? AND id = ?')
    .get(campaignId, sessionId);
  if (!session) {
    return { status: 404, body: { error: 'unknown session id' } };
  }

  const present = body.present;
  if (!isStringArray(present)) {
    return { status: 400, body: { error: 'present must be an array of strings' } };
  }

  const absent = body.absent;
  if (!isStringArray(absent)) {
    return { status: 400, body: { error: 'absent must be an array of strings' } };
  }

  db.prepare(
    `INSERT INTO campaign_session_attendance (campaign_id, session_id, present_json, absent_json)
     VALUES (?, ?, ?, ?)
     ON CONFLICT (campaign_id, session_id) DO UPDATE SET present_json = excluded.present_json, absent_json = excluded.absent_json`,
  ).run(campaignId, sessionId, JSON.stringify(present), JSON.stringify(absent));

  return {
    status: 200,
    body: {
      session_id: sessionId,
      present_count: present.length,
      absent_count: absent.length,
    },
  };
}

export function nextSession(campaignId: string): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const row = db
    .prepare(
      'SELECT id, starts_at, duration_minutes, agenda_json FROM campaign_sessions WHERE campaign_id = ? ORDER BY starts_at ASC LIMIT 1',
    )
    .get(campaignId) as CampaignSessionRow | undefined;

  if (!row) {
    return { status: 404, body: { error: 'no sessions scheduled' } };
  }

  const agenda = JSON.parse(row.agenda_json) as string[];

  return {
    status: 200,
    body: {
      id: row.id,
      starts_at: row.starts_at,
      agenda_count: agenda.length,
    },
  };
}
